File: android/app/src/main/java/com/sapiens/gagaodok/sync/SyncAccountTransitionCoordinator.kt
package com.sapiens.gagaodok.sync

data class SyncTransitionCandidate(
    val connection: SyncConnectionConfiguration,
    val secrets: SyncSecretBundle,
    val replicaData: ByteArray,
    val cursorData: ByteArray,
)

enum class SyncCommitBoundary {
    ROLLBACK_SECRET_SAVED,
    ROLLBACK_FILES_PREPARED,
    JOURNAL_COMMITTING,
    ACTIVE_SECRET_PROMOTED,
    ACTIVE_FILES_PROMOTED,
}

enum class SyncAccountTransitionState {
    IDLE, PREPARING, VERIFYING, BOOTSTRAPPING, READY_TO_COMMIT, COMMITTING,
    COMPLETED, CANCELLED, RECOVERABLE_ERROR, MANUAL_RECOVERY_REQUIRED,
}

enum class SyncAccountTransitionError {
    SYNC_ENABLED, OUTBOX_PENDING, ALREADY_SAME_ACCOUNT, CANDIDATE_UNVERIFIED,
    NO_ACTIVE_ACCOUNT, STORAGE_FAILED, MANUAL_RECOVERY_REQUIRED,
}

class SyncAccountTransitionException(val reason: SyncAccountTransitionError) : Exception(reason.name)

class SyncAccountTransitionCoordinator(
    private val vault: SyncSlottedSecretVault,
    private val connectionStore: SyncConnectionStateStore,
    private val files: SyncTransitionFiles,
    private val journal: SyncAccountTransitionJournal,
    private val outbox: SyncOutbox,
    private val nowMilliseconds: () -> Long = { System.currentTimeMillis() },
    private val afterBoundary: (SyncCommitBoundary) -> Unit = {},
) {
    var state: SyncAccountTransitionState = SyncAccountTransitionState.IDLE
        private set
    private var candidate: SyncTransitionCandidate? = null

    fun prepare(candidate: SyncTransitionCandidate) {
        val old = activeDisabledConnection()
        if (old.accountId == candidate.connection.accountId) fail(SyncAccountTransitionError.ALREADY_SAME_ACCOUNT)
        if (candidate.connection.enabled) fail(SyncAccountTransitionError.SYNC_ENABLED)
        state = SyncAccountTransitionState.PREPARING
        try {
            if (!vault.save(candidate.secrets, SyncSecretSlot.STAGING)) fail(SyncAccountTransitionError.STORAGE_FAILED)
            files.stage(connectionStore.encoded(candidate.connection), SyncTransitionFileRole.CONNECTION)
            files.stage(candidate.replicaData, SyncTransitionFileRole.REPLICA)
            files.stage(candidate.cursorData, SyncTransitionFileRole.CURSOR)
            journal.save(
                SyncAccountTransitionRecord(
                    stage = SyncTransitionStage.STAGED,
                    oldAccountId = old.accountId,
                    newAccountId = candidate.connection.accountId,
                    createdAtMilliseconds = nowMilliseconds(),
                )
            )
            this.candidate = candidate
            state = SyncAccountTransitionState.VERIFYING
        } catch (error: SyncAccountTransitionException) {
            cleanupQuietly()
            state = SyncAccountTransitionState.RECOVERABLE_ERROR
            throw error
        } catch (_: Exception) {
            cleanupQuietly()
            state = SyncAccountTransitionState.RECOVERABLE_ERROR
            fail(SyncAccountTransitionError.STORAGE_FAILED)
        }
    }

    fun markBootstrapVerified() {
        if (state == SyncAccountTransitionState.VERIFYING || state == SyncAccountTransitionState.BOOTSTRAPPING) {
            state = SyncAccountTransitionState.READY_TO_COMMIT
        }
    }

    fun commit() {
        val selected = candidate ?: fail(SyncAccountTransitionError.CANDIDATE_UNVERIFIED)
        if (state != SyncAccountTransitionState.READY_TO_COMMIT) fail(SyncAccountTransitionError.CANDIDATE_UNVERIFIED)
        val old = activeDisabledConnection()
        val activeSecrets = (vault.load(SyncSecretSlot.ACTIVE) as? SyncSecretLoadResult.Available)?.secrets
            ?: fail(SyncAccountTransitionError.STORAGE_FAILED)
        val stagingSecrets = (vault.load(SyncSecretSlot.STAGING) as? SyncSecretLoadResult.Available)?.secrets
            ?: fail(SyncAccountTransitionError.STORAGE_FAILED)
        if (stagingSecrets != selected.secrets) fail(SyncAccountTransitionError.STORAGE_FAILED)

        state = SyncAccountTransitionState.COMMITTING
        try {
            if (!vault.save(activeSecrets, SyncSecretSlot.ROLLBACK)) fail(SyncAccountTransitionError.STORAGE_FAILED)
            afterBoundary(SyncCommitBoundary.ROLLBACK_SECRET_SAVED)
            files.prepareRollback()
            afterBoundary(SyncCommitBoundary.ROLLBACK_FILES_PREPARED)
            journal.save(
                SyncAccountTransitionRecord(
                    stage = SyncTransitionStage.COMMITTING,
                    oldAccountId = old.accountId,
                    newAccountId = selected.connection.accountId,
                    createdAtMilliseconds = nowMilliseconds(),
                )
            )
            afterBoundary(SyncCommitBoundary.JOURNAL_COMMITTING)
            if (!vault.save(stagingSecrets, SyncSecretSlot.ACTIVE)) fail(SyncAccountTransitionError.STORAGE_FAILED)
            afterBoundary(SyncCommitBoundary.ACTIVE_SECRET_PROMOTED)
            files.promote()
            if (!files.removeActive(SyncTransitionFileRole.OUTBOX)) fail(SyncAccountTransitionError.STORAGE_FAILED)
            afterBoundary(SyncCommitBoundary.ACTIVE_FILES_PROMOTED)
            if (
                connectionStore.load() != SyncConnectionLoadResult.Available(selected.connection) ||
                selected.connection.enabled ||
                vault.load(SyncSecretSlot.ACTIVE) != SyncSecretLoadResult.Available(selected.secrets)
            ) fail(SyncAccountTransitionError.STORAGE_FAILED)
            cleanupTransient()
            candidate = null
            state = SyncAccountTransitionState.COMPLETED
        } catch (error: Exception) {
            state = SyncAccountTransitionState.RECOVERABLE_ERROR
            throw error
        }
    }

    fun cancel() {
        if (state == SyncAccountTransitionState.COMMITTING) fail(SyncAccountTransitionError.STORAGE_FAILED)
        cleanupTransient()
        candidate = null
        state = SyncAccountTransitionState.CANCELLED
    }

    fun recoverIfNeeded() {
        val record = journal.load()
        if (record == null) {
            cleanupTransient(removeJournal = false)
            state = SyncAccountTransitionState.IDLE
            return
        }
        when (record.operation to record.stage) {
            SyncTransitionOperation.SWITCH_ACCOUNT to SyncTransitionStage.STAGED -> {
                cleanupTransient(); candidate = null; state = SyncAccountTransitionState.IDLE
            }
            SyncTransitionOperation.SWITCH_ACCOUNT to SyncTransitionStage.COMMITTING -> recoverSwitch(record)
            SyncTransitionOperation.UNLINK to SyncTransitionStage.COMMITTING -> recoverUnlink(record)
            SyncTransitionOperation.UNLINK to SyncTransitionStage.STAGED -> {
                cleanupTransient(); state = SyncAccountTransitionState.IDLE
            }
        }
    }

    fun unlink() {
        val old = activeDisabledConnection()
        val activeSecrets = (vault.load(SyncSecretSlot.ACTIVE) as? SyncSecretLoadResult.Available)?.secrets
            ?: fail(SyncAccountTransitionError.NO_ACTIVE_ACCOUNT)
        try {
            if (!vault.save(activeSecrets, SyncSecretSlot.ROLLBACK)) fail(SyncAccountTransitionError.STORAGE_FAILED)
            files.prepareRollback()
            journal.save(
                SyncAccountTransitionRecord(
                    stage = SyncTransitionStage.COMMITTING,
                    operation = SyncTransitionOperation.UNLINK,
                    oldAccountId = old.accountId,
                    createdAtMilliseconds = nowMilliseconds(),
                )
            )
            if (!vault.remove(SyncSecretSlot.ACTIVE)) fail(SyncAccountTransitionError.STORAGE_FAILED)
            files.removeAllActive()
            if (connectionStore.load() != SyncConnectionLoadResult.Absent || vault.load(SyncSecretSlot.ACTIVE) != SyncSecretLoadResult.Absent) {
                fail(SyncAccountTransitionError.STORAGE_FAILED)
            }
            cleanupTransient()
            state = SyncAccountTransitionState.COMPLETED
        } catch (error: Exception) {
            state = SyncAccountTransitionState.RECOVERABLE_ERROR
            throw error
        }
    }

    private fun recoverSwitch(record: SyncAccountTransitionRecord) {
        if (activeMatches(record.newAccountId)) {
            cleanupTransient(); candidate = null; state = SyncAccountTransitionState.COMPLETED; return
        }
        val rollback = (vault.load(SyncSecretSlot.ROLLBACK) as? SyncSecretLoadResult.Available)?.secrets
            ?: manualRecovery()
        try {
            if (!vault.save(rollback, SyncSecretSlot.ACTIVE)) manualRecovery()
            files.restore()
            if (!activeMatches(record.oldAccountId)) manualRecovery()
            cleanupTransient(); candidate = null; state = SyncAccountTransitionState.IDLE
        } catch (_: Exception) { manualRecovery() }
    }

    private fun recoverUnlink(record: SyncAccountTransitionRecord) {
        if (connectionStore.load() == SyncConnectionLoadResult.Absent && vault.load(SyncSecretSlot.ACTIVE) == SyncSecretLoadResult.Absent) {
            cleanupTransient(); state = SyncAccountTransitionState.COMPLETED; return
        }
        val rollback = (vault.load(SyncSecretSlot.ROLLBACK) as? SyncSecretLoadResult.Available)?.secrets
            ?: manualRecovery()
        try {
            if (!vault.save(rollback, SyncSecretSlot.ACTIVE)) manualRecovery()
            files.restore()
            if (!activeMatches(record.oldAccountId)) manualRecovery()
            cleanupTransient(); state = SyncAccountTransitionState.IDLE
        } catch (_: Exception) { manualRecovery() }
    }

    private fun activeDisabledConnection(): SyncConnectionConfiguration {
        val connection = (connectionStore.load() as? SyncConnectionLoadResult.Available)?.configuration
            ?: fail(SyncAccountTransitionError.NO_ACTIVE_ACCOUNT)
        if (connection.enabled) fail(SyncAccountTransitionError.SYNC_ENABLED)
        if (outbox.pending().isNotEmpty()) fail(SyncAccountTransitionError.OUTBOX_PENDING)
        return connection
    }

    private fun activeMatches(accountId: String?): Boolean {
        val connection = (connectionStore.load() as? SyncConnectionLoadResult.Available)?.configuration ?: return false
        return accountId != null && connection.accountId == accountId && !connection.enabled &&
            vault.load(SyncSecretSlot.ACTIVE) is SyncSecretLoadResult.Available
    }

    private fun cleanupTransient(removeJournal: Boolean = true) {
        if (!vault.remove(SyncSecretSlot.STAGING) || !vault.remove(SyncSecretSlot.ROLLBACK)) {
            fail(SyncAccountTransitionError.STORAGE_FAILED)
        }
        files.discardTransient()
        if (removeJournal && !journal.remove()) fail(SyncAccountTransitionError.STORAGE_FAILED)
    }

    private fun cleanupQuietly() {
        runCatching { cleanupTransient() }
    }

    private fun manualRecovery(): Nothing {
        state = SyncAccountTransitionState.MANUAL_RECOVERY_REQUIRED
        fail(SyncAccountTransitionError.MANUAL_RECOVERY_REQUIRED)
    }

    private fun fail(reason: SyncAccountTransitionError): Nothing = throw SyncAccountTransitionException(reason)
}

File: Sources/KakaoSapiens/Services/SyncAccountTransitionCoordinator.swift
import Foundation

public struct SyncTransitionCandidate {
    public let connection: SyncConnectionConfiguration
    public let secrets: SyncSecretBundle
    public let replicaData: Data
    public let cursorData: Data

    public init(
        connection: SyncConnectionConfiguration,
        secrets: SyncSecretBundle,
        replicaData: Data,
        cursorData: Data
    ) {
        self.connection = connection
        self.secrets = secrets
        self.replicaData = replicaData
        self.cursorData = cursorData
    }
}

public enum SyncCommitBoundary: String, CaseIterable {
    case rollbackSecretSaved
    case rollbackFilesPrepared
    case journalCommitting
    case activeSecretPromoted
    case activeFilesPromoted
}

public enum SyncAccountTransitionState: Equatable {
    case idle
    case preparing
    case verifying
    case bootstrapping
    case readyToCommit
    case committing
    case completed
    case cancelled
    case recoverableError
    case manualRecoveryRequired
}

public enum SyncAccountTransitionError: Error, Equatable {
    case syncEnabled
    case outboxPending
    case alreadySameAccount
    case candidateUnverified
    case noActiveAccount
    case storageFailed
    case manualRecoveryRequired
}

public final class SyncAccountTransitionCoordinator {
    public private(set) var state: SyncAccountTransitionState = .idle

    private let vault: SyncSlottedSecretVault
    private let connectionStore: SyncConnectionStateStore
    private let files: SyncTransitionFiles
    private let journal: SyncAccountTransitionJournal
    private let outbox: SyncOutbox
    private let nowMilliseconds: () -> Int64
    private let afterBoundary: (SyncCommitBoundary) throws -> Void
    private var candidate: SyncTransitionCandidate?

    public init(
        vault: SyncSlottedSecretVault,
        connectionStore: SyncConnectionStateStore,
        files: SyncTransitionFiles,
        journal: SyncAccountTransitionJournal,
        outbox: SyncOutbox,
        nowMilliseconds: @escaping () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1_000) },
        afterBoundary: @escaping (SyncCommitBoundary) throws -> Void = { _ in }
    ) {
        self.vault = vault
        self.connectionStore = connectionStore
        self.files = files
        self.journal = journal
        self.outbox = outbox
        self.nowMilliseconds = nowMilliseconds
        self.afterBoundary = afterBoundary
    }

    public func prepare(candidate: SyncTransitionCandidate) throws {
        let old = try activeDisabledConnection()
        guard old.accountID != candidate.connection.accountID else {
            throw SyncAccountTransitionError.alreadySameAccount
        }
        guard !candidate.connection.enabled else { throw SyncAccountTransitionError.syncEnabled }

        state = .preparing
        do {
            try vault.save(candidate.secrets, slot: .staging)
            try files.stage(connectionStore.encoded(candidate.connection), role: .connection)
            try files.stage(candidate.replicaData, role: .replica)
            try files.stage(candidate.cursorData, role: .cursor)
            try journal.save(SyncAccountTransitionRecord(
                stage: .staged,
                oldAccountID: old.accountID,
                newAccountID: candidate.connection.accountID,
                createdAtMilliseconds: nowMilliseconds()
            ))
            self.candidate = candidate
            state = .verifying
        } catch {
            try? cleanupTransient()
            state = .recoverableError
            throw SyncAccountTransitionError.storageFailed
        }
    }

    public func markBootstrapVerified() {
        guard state == .verifying || state == .bootstrapping else { return }
        state = .readyToCommit
    }

    public func commit() throws {
        guard state == .readyToCommit, let candidate else {
            throw SyncAccountTransitionError.candidateUnverified
        }
        let old = try activeDisabledConnection()
        guard case .available(let activeSecrets) = vault.load(slot: .active),
              case .available(let stagingSecrets) = vault.load(slot: .staging),
              stagingSecrets == candidate.secrets else {
            throw SyncAccountTransitionError.storageFailed
        }

        state = .committing
        do {
            try vault.save(activeSecrets, slot: .rollback)
            try afterBoundary(.rollbackSecretSaved)
            try files.prepareRollback()
            try afterBoundary(.rollbackFilesPrepared)
            try journal.save(SyncAccountTransitionRecord(
                stage: .committing,
                oldAccountID: old.accountID,
                newAccountID: candidate.connection.accountID,
                createdAtMilliseconds: nowMilliseconds()
            ))
            try afterBoundary(.journalCommitting)
            try vault.save(stagingSecrets, slot: .active)
            try afterBoundary(.activeSecretPromoted)
            try files.promote()
            try files.removeActive(role: .outbox)
            try afterBoundary(.activeFilesPromoted)

            guard connectionStore.load() == .available(candidate.connection),
                  candidate.connection.enabled == false,
                  vault.load(slot: .active) == .available(candidate.secrets) else {
                throw SyncAccountTransitionError.storageFailed
            }
            try cleanupTransient()
            self.candidate = nil
            state = .completed
        } catch {
            state = .recoverableError
            throw error
        }
    }

    public func cancel() throws {
        guard state != .committing else { throw SyncAccountTransitionError.storageFailed }
        try cleanupTransient()
        candidate = nil
        state = .cancelled
    }

    public func recoverIfNeeded() throws {
        guard let record = try journal.load() else {
            try cleanupTransient(removeJournal: false)
            state = .idle
            return
        }

        switch (record.operation, record.stage) {
        case (.switchAccount, .staged):
            try cleanupTransient()
            candidate = nil
            state = .idle
        case (.switchAccount, .committing):
            if activeMatches(accountID: record.newAccountID) {
                try cleanupTransient()
                candidate = nil
                state = .completed
            } else if case .available(let rollback) = vault.load(slot: .rollback) {
                do {
                    try vault.save(rollback, slot: .active)
                    try files.restore()
                    guard activeMatches(accountID: record.oldAccountID) else {
                        throw SyncAccountTransitionError.manualRecoveryRequired
                    }
                    try cleanupTransient()
                    candidate = nil
                    state = .idle
                } catch {
                    state = .manualRecoveryRequired
                    throw SyncAccountTransitionError.manualRecoveryRequired
                }
            } else {
                state = .manualRecoveryRequired
                throw SyncAccountTransitionError.manualRecoveryRequired
            }
        case (.unlink, .committing):
            if connectionStore.load() == .absent && vault.load(slot: .active) == .absent {
                try cleanupTransient()
                state = .completed
            } else if case .available(let rollback) = vault.load(slot: .rollback) {
                try vault.save(rollback, slot: .active)
                try files.restore()
                guard activeMatches(accountID: record.oldAccountID) else {
                    state = .manualRecoveryRequired
                    throw SyncAccountTransitionError.manualRecoveryRequired
                }
                try cleanupTransient()
                state = .idle
            } else {
                state = .manualRecoveryRequired
                throw SyncAccountTransitionError.manualRecoveryRequired
            }
        case (.unlink, .staged):
            try cleanupTransient()
            state = .idle
        }
    }

    public func unlink() throws {
        let old = try activeDisabledConnection()
        guard case .available(let activeSecrets) = vault.load(slot: .active) else {
            throw SyncAccountTransitionError.noActiveAccount
        }
        do {
            try vault.save(activeSecrets, slot: .rollback)
            try files.prepareRollback()
            try journal.save(SyncAccountTransitionRecord(
                stage: .committing,
                operation: .unlink,
                oldAccountID: old.accountID,
                newAccountID: nil,
                createdAtMilliseconds: nowMilliseconds()
            ))
            vault.remove(slot: .active)
            try files.removeAllActive()
            guard connectionStore.load() == .absent, vault.load(slot: .active) == .absent else {
                throw SyncAccountTransitionError.storageFailed
            }
            try cleanupTransient()
            state = .completed
        } catch {
            state = .recoverableError
            throw error
        }
    }

    private func activeDisabledConnection() throws -> SyncConnectionConfiguration {
        guard case .available(let connection) = connectionStore.load() else {
            throw SyncAccountTransitionError.noActiveAccount
        }
        guard !connection.enabled else { throw SyncAccountTransitionError.syncEnabled }
        guard try outbox.pending().isEmpty else { throw SyncAccountTransitionError.outboxPending }
        return connection
    }

    private func activeMatches(accountID: String?) -> Bool {
        guard let accountID,
              case .available(let connection) = connectionStore.load(),
              case .available = vault.load(slot: .active) else { return false }
        return connection.accountID == accountID && connection.enabled == false
    }

    private func cleanupTransient(removeJournal: Bool = true) throws {
        vault.remove(slot: .staging)
        vault.remove(slot: .rollback)
        try files.discardTransient()
        if removeJournal { try journal.remove() }
    }
}
